manifests/administration.pp

```diff
@@ -30,14 +30,43 @@
       value  => '1',
     }
   }
+  file { '/etc/systemd/system/disable_thp.service':
+    ensure  => file,
+    mode    => '0644',
+    owner   => 'root',
+    group   => 'root',
+    content => file('redis/service_files/disable_thp.service'),
+  }
 
-  if $disable_thp {
-    exec { 'Disable Hugepages':
-      command => 'echo never > /sys/kernel/mm/transparent_hugepage/enabled',
-      path    => ['/sbin', '/usr/sbin', '/bin', '/usr/bin'],
-      onlyif  => 'test -f /sys/kernel/mm/transparent_hugepage/enabled',
-      unless  => 'cat /sys/kernel/mm/transparent_hugepage/enabled | grep "\[never\]"',
-    }
+  # Only necessary for Puppet < 6.1.0,
+  # See https://github.com/puppetlabs/puppet/commit/f8d5c60ddb130c6429ff12736bfdb4ae669a9fd4
+  if versioncmp($facts['puppetversion'],'6.1.0') < 0 {
+    include systemd::systemctl::daemon_reload
+    File['/etc/systemd/system/disable_thp.service'] ~> Class['systemd::systemctl::daemon_reload']
+  }
+
+  $hugeadm_package = $facts['os']['family'] ? {
+    'RedHat' => 'libhugetlbfs-utils',
+    'Debian' => 'libhugetlbfs',
+    'Archlinux' => 'libhugetlbfs',
+    default  => 'hugepages',
+  }
+
+  package { $hugeadm_package:
+    ensure => 'present',
+  }
+
+  service { 'disable_thp':
+    ensure    => false,
+    enable    => $disable_thp,
+    subscribe => File['/etc/systemd/system/disable_thp.service'],
+  }
+
+  exec { 'systemd run_once disable_thp':
```

Comment:
AFAICT this exec will *always* run, even if `$disable_thp` is false. That's probably not supposed to happen?

In general, it seems a bit hacky to make a one-off subscribe to a file change. You might want to consider writing a quick fact that checks the status and ensures thp is disabled.

Reply:
Or, as I just checked up on docs, add a timer unit with OnBootSec=0 as a trigger, that *should* fire the moment you start the timer.

Reply:
@towo  It was indeed hacked to ensure THP will be disabled, before reboot for initial setup of host.
I have introduced timer as you have suggested.